crates/gitbutler-repo/src/staging.rs

```diff
@@ -0,0 +1,102 @@
+use std::{collections::HashMap, path::PathBuf};
+
+use anyhow::{anyhow, Result};
+use git2::{ApplyLocation, ApplyOptions, Repository};
+use gitbutler_command_context::CommandContext;
+use gitbutler_diff::{ChangeType, FileDiff, GitHunk, Hunk};
+use serde::{Deserialize, Serialize};
+
+fn stage_hunk(ctx: &CommandContext, path: &PathBuf, hunks: &Vec<GitHunk>) -> Result<()> {
+    let repo = ctx.repo();
+    let mut index = repo.index()?;
+    if hunks.iter().any(|h| h.change_type == ChangeType::Untracked) {
+        index.add_path(path)?;
+        index.write()?;
+        return Ok(());
+    }
+
+    let mut apply_opts = ApplyOptions::new();
+    apply_opts.hunk_callback(|cb_hunk| {
+        cb_hunk.map_or(false, |cb_hunk| {
+            for hunk in hunks {
+                if hunk.new_start == cb_hunk.new_start()
+                    && hunk.new_start + hunk.new_lines == cb_hunk.new_start() + cb_hunk.new_lines()
+                {
+                    return true;
+                }
+            }
+            false
+        })
+    });
+
+    let diff = diff_workdir_to_index(repo, path)?;
+    repo.apply(&diff, ApplyLocation::Index, Some(&mut apply_opts))?;
+    Ok(())
+}
+
+pub fn stage(ctx: &CommandContext, files_to_stage: &Vec<(PathBuf, Vec<GitHunk>)>) -> Result<()> {
+    for (path_to_stage, hunks_to_stage) in files_to_stage {
+        stage_hunk(ctx, path_to_stage, hunks_to_stage)?;
+    }
+    Ok(())
+}
+
+fn diff_workdir_to_index<'a>(repo: &'a Repository, path: &PathBuf) -> Result<git2::Diff<'a>> {
+    let index = repo.index()?;
+    let mut diff_opts = git2::DiffOptions::new();
+    diff_opts
+        .recurse_untracked_dirs(true)
+        .include_untracked(true)
+        .show_binary(true)
+        .show_untracked_content(true)
+        .ignore_submodules(true)
+        .context_lines(3)
+        .pathspec(path);
+    Ok(repo.diff_index_to_workdir(Some(&index), Some(&mut diff_opts))?)
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize, Default)]
+pub struct HunkId {
+    pub start: u32,
+    pub end: u32,
+}
```

Comment:
@krlvi could you help me think this through. Originally we relied on `OwnershipClaim`, but it's located in `gitbutler-stack`. We could use `Hunk` from `gitbutler-diff`, but unsure if we want to involve the hunk hash in staging.

Reply:
What is your rationale against using `Hunk` from `gitbutler-diff`? Creating this additional one here may create extra work for us down the line. Since you are doing the From conversion anyways, there is a dependency already

Reply:
We can use that one, I just don't like that it has the extra hash.. so in tests about staging you have to construct something like `Hunk { start: (), end: (), hash: None }`. Would you prefer I switch to that?

Reply:
I've switched it over to `Hunk` now.